File: Review-15-Python-Strategy-Pattern/Example-3/tictactoe/strategy.py
from dataclasses import dataclass
from typing import Never, Protocol


class Strategy(Protocol):
    def next_move(self) -> int:
        """
        Retrieve the next move.
        """


PROMPT_MSG: str = "Enter your desired move (1-9): "


@dataclass
class KeyboardStrategy:
    _name: str = ""

    def next_move(self) -> int:
        choice = int(input(f"{self._name}, {PROMPT_MSG}"))

        return choice


class PredefinedMoves:
    def __init__(self, *, moves: list[int]) -> None:
        self.my_moves = moves
        self.__move_idx = 0

    def next_move(self) -> int:
        my_next_move = self.my_moves[self.__move_idx]

        self.__move_idx += 1

        return my_next_move

File: Review-04-CPP-Unit-Testing/Example-1/source/Strategy.cpp
#include <iostream>
#include <string>

#include "Strategy.h"

const std::string KeyboardStrategy::PROMPT_MSG = "Enter your desired move (1-9): ";

KeyboardStrategy::KeyboardStrategy(std::string name)
    :_name(name)
{
}

KeyboardStrategy::~KeyboardStrategy()
{
}

int KeyboardStrategy::nextMove()
{
    int choice;

    std::cout << (this->_name + ", " + KeyboardStrategy::PROMPT_MSG);
    std::cin >> choice;

    return choice;
}
